### PR TITLE
Add EV, R2, MSE and MAE to metadata 

### LINKED ISSUE #360: Support other metrics for hyperparameter tuning in Katib
Currently, in the yaml file used to generate a Katib workflow, we support
- explained-variance_mean as the _objectivevaluename_  and _metricsnames_
- explained-variance_std, explained-variance_max and explained-variance_mean as _metricsnames_

These metrics are aggregates over all sensors and validation splits.

1.  
It would be useful to support other metrics such as:
- mean squared error
- R2 score
- mean absolute error

2. 
It would also be useful to output metrics for each sensor and each validation split in addition to aggregated values.

### PR DESCRIPTION
**Pull request by Natalie-Caruana**
This will close #360 and #401 

Add EV, R2, MSE and MAE 
 - for each sensor and each validation split
-  aggregated (mean) scores across sensors for each validation split
-  aggregated scores across sensors and validation splits

to metadata.

### CODE REVIEW

File: gordo_components/builder/build_model.py [hunk 1]
```diff
@@ -75,16 +77,37 @@ def build_model(
...
+                "min": cv["test_score"].min(),
+                "raw-scores": cv["test_score"].tolist(),
             }
         }
+
+        # adding loss function to scores if Gordo model is a KerasBaseEstimator
+        gordo_model = _get_final_gordo_base_step(model)
+        if issubclass(gordo_model.__class__, KerasBaseEstimator):
+            val_losses = np.array([])
+            for mod in cv["estimator"]:
```
Comment (epa095) [inline gordo_components/builder/build_model.py]: So this loops over each of the trained models? You want the loss for all of them? And does it make sense to have the mean/std, etc? If those values are not meaningful for anything really then it is better to not add them.

File: gordo_components/builder/build_model.py [hunk 2]
```diff
@@ -75,16 +77,37 @@ def build_model(
...
+                gordo_model_cv = _get_final_gordo_base_step(mod)
+                val_loss = gordo_model_cv.get_metadata()["history"]["loss"][0]
```
Comment (epa095) [inline gordo_components/builder/build_model.py]: What does the `[0]` index into? The epochs?`If so, do you only want the loss for the first epoch?

File: gordo_components/builder/build_model.py [hunk 3]
```diff
@@ -75,16 +77,37 @@ def build_model(
...
+                val_losses = np.append(val_losses, val_loss)
+            scores["cv-loss"] = {
```
Comment (epa095) [inline gordo_components/builder/build_model.py]: Seems like this assignment should be after the for-loop.

File: gordo_components/builder/build_model.py [hunk 4]
```diff
@@ -75,16 +77,37 @@ def build_model(
...
+                "mean": val_losses.mean(),
+                "std": val_losses.std(),
+                "max": val_losses.max(),
+                "min": val_losses.min(),
+                "raw-scores": val_losses.tolist(),
```
Comment (epa095) [inline gordo_components/builder/build_model.py]: This can become quite big, no?

File: gordo_components/builder/build_model.py (hunk 2 above)
Comment (Natalie-Caruana) [reply gordo_components/builder/build_model.py]: That's a mistake...should read [-1] so the loss from last epoch is extracted.

File: gordo_components/builder/build_model.py (hunk 1 above)
Comment (Natalie-Caruana) [reply gordo_components/builder/build_model.py]: i think it makes sense to have the mean/std/min/max especially if we decide to increase the number of splits as it would give us an idea of the variability of the validation losses for the different splits.

**Comment by ryanjdillon** [comment]
Is this resolving #401?

File: gordo_components/builder/build_model.py [hunk 5]
```diff
@@ -89,30 +94,66 @@ def build_model(
...
         logger.debug("Starting cross validation")
         start = time.time()
-
-        scores: Dict[str, Any]
-
+        scores: Dict[str, Any] = dict()
         if hasattr(model, "predict"):
-
-            cv_scores = cross_val_score(
+            metrics_list = [
```
Comment (epa095) [inline gordo_components/builder/build_model.py]: Can maybe this whole thing be extracted to a function?

File: gordo_components/builder/build_model.py [hunk 6]
```diff
@@ -89,30 +94,66 @@ def build_model(
...
+                explained_variance_score,
+                r2_score,
+                mean_squared_error,
+                mean_absolute_error,
+            ]
+            metrics_dict = {}
+            for metric in metrics_list:
+                for index, col in enumerate(y.columns):
+
+                    def my_custom_score(y_true, y_pred, metric=metric, index=index):
```
Comment (epa095) [inline gordo_components/builder/build_model.py]: And this thing, can it be moved out, and given a better name? And maybe even a docstring. But it can be hidden (starting with a underscore)

Comment (Natalie-Caruana) [reply gordo_components/builder/build_model.py]: I tried to remove this function out of the for loops and create a wrapper, something like this:

```            def my_custom_score(y_true, y_pred, metric=r2_score, index=0):
                score = metric(y_true.values[:, index], y_pred[:, index])
                return score
```
and then in the for loop use a lambda function:
```    for metric in metrics_list:
        for index, col in enumerate(y.columns):
            metric_str = metric.__name__
            metrics_dict.update(
                {
                    metric_str.replace("_", "-")
                    + f'-{col.replace(" ", "-")}': make_scorer(
                        metric_wrapper(lambda y_true,y_pred: my_custom_score(y_true,y_pred,metric=metric,index=index)
                    )
                }
            )
```
but im getting the same results for different metrics/indexes...it looks like (since the lambda function depends only on y_true and y_pred that the metric and index are not being iterated...

File: gordo_components/cli/cli.py [hunk 7]
```diff
@@ -345,16 +345,21 @@ def get_all_score_strings(metadata):
...
     ...     "model": {
     ...         "cross-validation": {
-    ...             "scores": {"explained variance": {"min": 0, "max": 2}}
-    ...         }
+    ...             "scores": {"explained variance tag 0": {"fold_1": 0, "fold_2": 0.9, "fold_3": 0.1,"min": 0.1, "max": 0.9, "mean": 1/3},
+    ...                        "explained variance tag 1": {"fold_1": 0.2, "fold_2": 0.3, "fold_3": 0.6, "min": 0.2, "max": 0.6, "mean": 1.1/3},
+    ...                        "explained variance mean" : {"min": 0.2/2, "max": 1.2/2, "mean": (1/3)*(0.2/2 + 1.2/2 + 0.7/2)},
+    ...                        "r2 score tag 0" : {"fold_1": 0.8, "fold_2": 0.5, "fold_3": 0.6, "min": 0.5, "max": 0.8, "mean": 1.9/3},
+    ...                        "r2 score tag 1" : {"fold_1": 0.4, "fold_2": 0.3, "fold_3": 0.5, "min": 0.3, "max": 0.5, "mean": 1.2/3},
+    ...                        "r2 score mean"  : {"min": 0.8/2,"max": 1.2/2, "mean": (1/3)*(1.2/2 + 0.8/2 + 1.1/2)}
```
[the remark this answers is not in the document]
Comment (Natalie-Caruana) [reply gordo_components/cli/cli.py]: As discussed my aim was to show what it was actually being computed...that it basically averages both over folds and tags.  But i will explain this in a comment as suggested

File: gordo_components/builder/build_model.py [hunk 8]
```diff
@@ -85,34 +90,48 @@ def build_model(
...
+                    "fold-mean": cv[f"test_{key}"].mean(),
+                    "fold-std": cv[f"test_{key}"].std(),
+                    "fold-max": cv[f"test_{key}"].max(),
+                    "fold-min": cv[f"test_{key}"].min(),
                 }
-            }
+                val.update(
+                    {
+                        f"fold-{i + 1}": raw_value
+                        for i, raw_value in enumerate(cv[f"test_{key}"].tolist())
```
Comment (milesgranger) [inline gordo_components/builder/build_model.py]: instead of all these `f"test_{key}"` , maybe create a variable, or update `key` in a map during the for loop. ie.
```python
for key in map(lambda k: f"test_{k}", metrics_dict):
```

File: gordo_components/builder/build_model.py [hunk 9]
```diff
@@ -85,34 +90,48 @@ def build_model(
...
+                    }
+                )
+                scores.update({key: val})
+
         else:
-            logger.debug("Unable to score model, has no attribute 'score'.")
+            logger.debug("Unable to score model, has no attribute 'predict'.")
```
Comment (milesgranger) [inline gordo_components/builder/build_model.py:131]: Nice catchy! :+1:

File: gordo_components/builder/build_model.py [hunk 10]
```diff
@@ -150,10 +169,62 @@ def build_model(
...
 
     metadata["model"].update(_get_metadata(model))
-
     return model, metadata
 
 
+def get_metrics_dict(metrics_list, y):
+    """
+        Given a list of metrics that accept a true_y and pred_y as inputs this returns
+        a dictionary with keys in the form '{score}-{tag_name}' for each given target tag
```
Comment (milesgranger) [inline gordo_components/builder/build_model.py]: _very_ mild, but you have an extra indent here in the docs.

Also, as we talked, maybe it is more generic to call things `feature(s)` rather than tags? I'm not terribly concerned about this though...

File: gordo_components/builder/build_model.py [hunk 11]
```diff
@@ -150,10 +169,62 @@ def build_model(
...
+            Target data
+
+
+        Returns
+        -------
+            dict
+        """
+
+    def _score_factory(metric=r2_score, index=0):
+        def _score_per_tag(y_true, y_pred):
```
Comment (milesgranger) [inline gordo_components/builder/build_model.py:196]: Nice and clean!

File: gordo_components/builder/build_model.py [hunk 12]
```diff
@@ -150,10 +169,62 @@ def build_model(
...
+            # This function extracts the score for each given target_tag to
+            # use as scoring argument in sklearn cross_validate, as the scoring
+            # must return a single value.
+            return metric(y_true.values[:, index], y_pred[:, index])
```
Comment (milesgranger) [inline gordo_components/builder/build_model.py]: Seems like we're making some fairly inflexible assumptions here.
1. That `y_true` _will_ be a dataframe
2. That `y_pred` _will_ be a numpy array

Perhaps we can accept either for both and do some `hasattr(..., 'values')` or something similar to be a bit more flexible?

File: gordo_components/builder/build_model.py [hunk 13]
```diff
@@ -150,10 +169,62 @@ def build_model(
...
+                    + f'-{col.replace(" ", "-")}': make_scorer(
+                        metric_wrapper(_score_factory(metric=metric, index=index))
+                    )
+                }
+            )
+
+        metrics_dict.update(
+            {
+                metric_str.replace("_", "-")
+                + f"-tag-mean": make_scorer(metric_wrapper(metric))
```
Comment (milesgranger) [inline gordo_components/builder/build_model.py]: here i think we talked about removing 'tag' maybe?

File: tests/gordo_components/builder/test_builder.py [hunk 14]
```diff
@@ -50,7 +50,39 @@ def metadata_check(metadata, check_history):
...
+            for tag in metadata["dataset"]["tag_list"]
+        ]
+        assert (
+            "explained-variance-score-tag-mean"
+            in metadata["model"]["cross-validation"]["scores"]
+        )
+        assert [
+            f"mean-squared-error-{tag.name.replace(' ','-')}"
+            in metadata["model"]["cross-validation"]["scores"]
+            for tag in metadata["dataset"]["tag_list"]
```
Comment (milesgranger) [inline tests/gordo_components/builder/test_builder.py]: seems like lots of `metadata["dataset"]["tag_list"]` followed by `tag.name.replace(...)` maybe at the start of this conditional, do the preprocessing into `tag_list` and probably `scores` from `metadata["model"]["cross-validation"]["scores"]` 

That might actually save you a bunch of lines from the formatting.

File: tests/gordo_components/builder/test_builder.py [hunk 15]
```diff
@@ -262,6 +294,174 @@ def test_get_metadata_helper(model: BaseEstimator, expect_empty_dict: bool):
...
+    model3, metadata3 = build_model(
+        name="model-name",
+        model_config=model_config3,
+        data_config=data_config,
+        metadata={},
+    )
+
+    metadata_check(metadata1, False)
+    metadata_check(metadata2, False)
+    metadata_check(metadata3, False)
```
Comment (milesgranger) [inline tests/gordo_components/builder/test_builder.py]: This test, specifically the `raw_model` configs, should be parameterized `pytest.mark.parameterize`

File: tests/gordo_components/builder/test_builder.py [hunk 16]
```diff
@@ -262,6 +294,174 @@ def test_get_metadata_helper(model: BaseEstimator, expect_empty_dict: bool):
...
+
+    assert (
+        metadata["model"]["cross-validation"]["scores"]["mean-absolute-error-Tag-1"][
+            "fold-mean"
+        ]
+        + metadata["model"]["cross-validation"]["scores"]["mean-absolute-error-Tag-2"][
+            "fold-mean"
+        ]
+    ) / 2 == pytest.approx(
+        metadata["model"]["cross-validation"]["scores"]["mean-absolute-error-tag-mean"][
```
Comment (milesgranger) [inline tests/gordo_components/builder/test_builder.py]: again for readability, creating a variable `scores = metadata["model"]["cross-validation"]["scores"]` would save on readability in this test.

**Review by milesgranger — COMMENTED** [review]
Sorry for the comments, mostly small things I think. Looks nice in general. :+1:

File: gordo_components/builder/build_model.py [hunk 17]
```diff
@@ -217,10 +219,7 @@ def _score_per_tag(y_true, y_pred):
             )
 
         metrics_dict.update(
-            {
-                metric_str.replace("_", "-")
-                + f"-tag-mean": make_scorer(metric_wrapper(metric))
-            }
+            {metric_str.replace("_", "-"): make_scorer(metric_wrapper(metric))}
```
Comment (milesgranger) [inline gordo_components/builder/build_model.py]: I think everywhere you use `metric_str` you do this `.replace(...)` maybe just do that at the assignment 
`metric_str = metric.__name__.replace('_', '-')`

File: tests/gordo_components/builder/test_builder.py [hunk 18]
```diff
@@ -50,39 +50,22 @@ def metadata_check(metadata, check_history):
...
-        assert [
-            f"mean-absolute-error-{tag.name.replace(' ','-')}"
-            in metadata["model"]["cross-validation"]["scores"]
-            for tag in metadata["dataset"]["tag_list"]
-        ]
-        assert (
-            "mean-absolute-error-tag-mean"
-            in metadata["model"]["cross-validation"]["scores"]
-        )
+        all_scores_list = [
```
Comment (milesgranger) [inline tests/gordo_components/builder/test_builder.py:62]: Nice! :100:

File: gordo_components/builder/build_model.py [hunk 19]
```diff
@@ -150,10 +166,64 @@ def build_model(
     }
 
     metadata["model"].update(_get_metadata(model))
-
     return model, metadata
 
 
+def get_metrics_dict(metrics_list, y):
```
Comment (milesgranger) [inline gordo_components/builder/build_model.py]: type defs

File: gordo_components/builder/build_model.py [hunk 20]
```diff
@@ -150,10 +166,63 @@ def build_model(
...
+
+    def _score_factory(metric=r2_score, index=0):
+        def _score_per_tag(y_true, y_pred):
+            # This function extracts the score for each given target_tag to
+            # use as scoring argument in sklearn cross_validate, as the scoring
+            # must return a single value.
+            if hasattr(y_true, "values"):
+                y_true = y_true.values
+            if hasattr(y_pred, "values"):
+                y_true = y_pred.values
```
Comment (milesgranger) [inline gordo_components/builder/build_model.py]: Here's a fun bug...

Comment (Natalie-Caruana) [reply gordo_components/builder/build_model.py]: grrrrr :angry:

Comment (Natalie-Caruana) [reply gordo_components/builder/build_model.py]: angry at me hehe

**Comment by milesgranger** [comment]
@Natalie-Caruana 
I re-ran the workflow you submitted earlier, as all the models were cached because their config didn't actually change. Anyhow, I can confirm it works! Nice job!

image
